serrano/resources/preview.py

```diff
@@ -100,31 +105,34 @@ def get(self, request):
             iterable = processor.get_iterable(queryset=queryset,
                                               request=request)
 
-            exported = exporter.read(iterable,
-                                     request=request,
-                                     offset=offset,
-                                     limit=limit)
+            rows = exporter.manual_read(iterable,
+                                        request=request,
+                                        offset=offset,
+                                        limit=limit)
         else:
             iterable = processor.get_iterable(queryset=queryset,
                                               request=request,
                                               limit=limit,
                                               offset=offset)
 
-            exported = exporter.read(iterable, request=request)
+            # Select the requested reader
+            reader = params['reader']
 
-        for row in exported:
-            pk = None
-            values = []
+            if reader == 'threaded':
+                method = exporter.threaded_read
+            elif reader == 'cached':
+                method = exporter.cached_read
+            elif reader == 'cached_threaded':
+                method = exporter.cached_threaded_read
+            else:
+                method = exporter.read
 
-            for i, output in enumerate(row):
-                if i == 0:
-                    pk = output[pk_name]
-                else:
-                    values.extend(output.values())
+            rows = method(iterable, request=request)
 
+        for row in rows:
             objects.append({
-                'pk': pk,
-                'values': values,
+                'pk': row[0],
```

Comment:
Same concern as above. Where is the pk being first guaranteed? I trust it is, just looking for the place where it is actually happening.